neqo-qpack/src/encoder.rs

```diff
@@ -525,9 +546,13 @@ mod tests {
         let mut encoder = connect(false);
 
         // insert "content-length: 1234
-        let res = encoder
-            .encoder
-            .insert_with_name_literal(HEADER_CONTENT_LENGTH, VALUE_1);
+        let res = encoder.encoder.insert(
+            &mut encoder.conn,
+            &EncoderInstruction::InsertWithNameLiteral {
+                name: HEADER_CONTENT_LENGTH.to_vec(),
```

Comment:
All these cases of `to_vec()` make me think that perhaps we need these to be declared as `Vec<u8>` instead.

But I wonder if instead we can avoid that entirely by changing `EncoderInstruction` to use references.  It's a little bit fiddly, but marshaling is easy as we don't really hold these instructions for very long.

```rust
pub enum EncoderInstruction<'a> {
    Capacity { value: u64 },
    InsertWithNameRefStatic { index: u64, value: &'a [u8] },
    InsertWithNameRefDynamic { index: u64, value: &'a [u8] },
    InsertWithNameLiteral { name: &'a [u8], value: &'a [u8] },
    Duplicate { index: u64 },
    NoInstruction,
}
```
The real tricky part is the decoder.  But by moving the reader that is attached to each `EncoderInstructionReaderState` enum arm to `EncoderInstructionReader` you could return a borrow of those underlying buffers from `EncoderInstructionReader::read_instructions`.  That avoids clones on decode, and only a little buffering.

That is enough work that this might be worth a separate issue.

Reply:
I know about this problem. And I will salve it in the next issue. It is not as simple as this, it is more complicated. 
enum EncoderInstruction are used for decoding instruction as well. In case of decoding I want them to own, hold an array (Vec<u8>)
I do not want 2 enums for the same thing. I will make them enum EncoderInstruction<T>.
This will need to be a separate issue.